File: src/main/java/com/rafaelfiume/tictactoe/TicTacToePrinter.java
package com.rafaelfiume.tictactoe;

import static java.lang.String.format;
import static java.lang.System.lineSeparator;

public class TicTacToePrinter {

    private static final String BLANK = " ";
    private static final String CHOOSE_POSITION = "Choose Position:";

    private final Board snapshot;

    private final PrinterState printerState;

    private final String boardTemplate = "" +
            " %s | %s | %s " + lineSeparator() +
            "---+---+---"    + lineSeparator() +
            " %s | %s | %s " + lineSeparator() +
            "---+---+---"    + lineSeparator() +
            " %s | %s | %s ";

    public TicTacToePrinter(Board currentSnapshot) {
        this.snapshot = currentSnapshot;
        this.printerState = printerStateFrom(currentSnapshot);
    }

    public String gameDescription() {
        return printerState.gameDescription();
    }

    public String board() {
        return format(boardTemplate,
                topLeft(), topCenter(), topRight(),
                midLeft(), center(), midRight(),
                bottomLeft(), bottomCenter(), bottomRight());
    }

    public String gameStatus() {
        return printerState.gameStatus();
    }

    private PrinterState printerStateFrom(Board currentSnapshot) {
        if (currentSnapshot.gameHasNotStarted())     return new EmptyBoardState();
        if (currentSnapshot.gameIsRunning())         return new PlayerTurnState(currentSnapshot.currentPlayer());
        if (currentSnapshot.gameIsOverWithAWinner()) return new PlayerWonState(currentSnapshot.winner());
        if (currentSnapshot.gameIsOverWithADraw())   return new DrawState();

        throw new IllegalStateException("Unknown board state");
    }

    private interface PrinterState {
        String gameDescription();
        String gameStatus();
    }

    private class EmptyBoardState implements PrinterState {

        @Override
        public String gameDescription() {
            return "Game Board Creation...";
        }

        @Override
        public String gameStatus() {
            return "Board Created." + lineSeparator() +
                    "The game will start with Player X" + lineSeparator() +
                    CHOOSE_POSITION;
        }
    }

    private class PlayerTurnState implements PrinterState {

        private final Player player;

        PlayerTurnState(Player player) {
            this.player = player;
        }

        @Override
        public String gameDescription() {
            return format("Player %s:", player);
        }

        @Override
        public String gameStatus() {
            return CHOOSE_POSITION;
        }
    }

    private class PlayerWonState implements PrinterState {

        private final Player winner;

        PlayerWonState(Player winner) {
            this.winner = winner;
        }

        @Override
        public String gameDescription() {
            return format("Player %s:", winner);
        }

        @Override
        public String gameStatus() {
            return format("PLAYER %s WON!", winner);
        }
    }

    private class DrawState implements PrinterState {

        @Override
        public String gameDescription() {
            return "No More Turns Left :-)";
        }

        @Override
        public String gameStatus() {
            return "GAME ENDS WITH A DRAW!";
        }
    }

    //
    // Boring methods
    //

    private String topLeft() {      return markAt(snapshot.topLeft()); }
    private String topCenter() {    return markAt(snapshot.topCenter()); }
    private String topRight() {     return markAt(snapshot.topRight()); }
    private String midLeft() {      return markAt(snapshot.midLeft()); }
    private String center() {       return markAt(snapshot.center()); }
    private String midRight() {     return markAt(snapshot.midRight()); }
    private String bottomLeft() {   return markAt(snapshot.bottomLeft()); }
    private String bottomCenter() { return markAt(snapshot.bottomCenter()); }
    private String bottomRight() {  return markAt(snapshot.bottomRight());}

    private String markAt(char c) {
        return (isEmpty(c)) ? BLANK : Character.toString(c);
    }

    private boolean isEmpty(char c) {
        return !Character.isLetter(c);
    }

}

File: src/main/java/com/rafaelfiume/tictactoe/TicTacToePrinter.kt
package com.rafaelfiume.tictactoe

import java.lang.String.format
import java.lang.System.lineSeparator

class TicTacToePrinter(private val snapshot: Game) {

    private val printerState: PrinterState

    private val boardTemplate = "" +
            " %s | %s | %s " + lineSeparator() +
            "---+---+---"    + lineSeparator() +
            " %s | %s | %s " + lineSeparator() +
            "---+---+---"    + lineSeparator() +
            " %s | %s | %s "

    init {
        this.printerState = printerStateFrom(snapshot)
    }

    fun gameDescription() = printerState.gameDescription()

    fun board(): String {
        return format(boardTemplate,
                topLeft()   , topCenter()   , topRight(),
                midLeft()   , center()      , midRight(),
                bottomLeft(), bottomCenter(), bottomRight())
    }

    fun gameStatus() = printerState.gameStatus()

    private fun printerStateFrom(game: Game): PrinterState {
        if (game.gameHasNotStarted()) return EmptyBoardState()
        if (game.gameIsOn())          return PlayerTurnState(game.currentPlayer())
        if (game.hasWinner())         return PlayerWonState(game.winner())
        if (game.hasDraw())           return DrawState()

        throw IllegalStateException("Unknown game state")
    }

    private interface PrinterState {
        fun gameDescription(): String
        fun gameStatus(): String
    }

    private inner class EmptyBoardState : PrinterState {

        override fun gameDescription() = "Game Board Creation..."
        override fun gameStatus() =
             "Board Created." + lineSeparator() +
             "The game will start with Player X" + lineSeparator() +
             CHOOSE_POSITION
    }

    private inner class PlayerTurnState internal constructor(private val player: Player) : PrinterState {

        override fun gameDescription() = "Player $player:"
        override fun gameStatus() = CHOOSE_POSITION
    }

    private inner class PlayerWonState internal constructor(private val winner: Player) : PrinterState {

        override fun gameDescription() = "Player $winner:"
        override fun gameStatus() = "PLAYER $winner WON!"
    }

    private inner class DrawState : PrinterState {

        override fun gameDescription() = "No More Turns Left :-)"
        override fun gameStatus() = "GAME ENDS WITH A DRAW!"
    }

    //
    // Boring methods
    //

    private fun topLeft()      = markedAt(snapshot.topLeft())
    private fun topCenter()    = markedAt(snapshot.topCenter())
    private fun topRight()     = markedAt(snapshot.topRight())
    private fun midLeft()      = markedAt(snapshot.midLeft())
    private fun center()       = markedAt(snapshot.center())
    private fun midRight()     = markedAt(snapshot.midRight())
    private fun bottomLeft()   = markedAt(snapshot.bottomLeft())
    private fun bottomCenter() = markedAt(snapshot.bottomCenter())
    private fun bottomRight()  = markedAt(snapshot.bottomRight())

    private fun markedAt(c: Char) = if (isEmpty(c)) BLANK else Character.toString(c)

    private fun isEmpty(c: Char) = !Character.isLetter(c)

    companion object {

        private val BLANK = " "
        private val CHOOSE_POSITION = "Choose Position:"
    }

}
